### PR TITLE
enable multi-apps for SSC

### PR DESCRIPTION
**Pull request by shevateng0**
Need the ssc engine to run multiple applications (>2). However, the original program can only communicate between the intra communicators of two applications. I add some codes to allow the commnunication among the intra communicators of multiple applications.

### CODE REVIEW

File: source/adios2/engine/ssc/SscReader.cpp [hunk 1]
```diff
@@ -117,6 +117,235 @@ void SscReader::SyncMpiPattern()
...
+        {
+            ilen = m_ReaderGlobalMpiInfo[i].size();
+        }
+        MPI_Bcast(&ilen, 1, MPI_INT, 0, MPI_COMM_WORLD);
+        m_ReaderGlobalMpiInfo[i].resize(ilen);
+        MPI_Bcast(m_ReaderGlobalMpiInfo[i].data(), ilen, MPI_INT, 0, MPI_COMM_WORLD); // Broadcast readerinfo size
+    }
+
+
+    std::cout << "WorldRank " << m_WorldRank << std::endl;
```
Comment (JasonRuonanWang) [inline source/adios2/engine/ssc/SscReader.cpp]: Could you remove these printing code when it's done?

File: source/adios2/engine/ssc/SscReader.cpp [hunk 2]
```diff
@@ -117,6 +117,235 @@ void SscReader::SyncMpiPattern()
...
+    std::vector<std::vector<int>> m_ReaderGlobalMpiInfo;
+
+
+    // Process m_WorldRank == 0 to gather all the local rank m_WriterRank, and find out all the m_WriterRank == 0
+    if (m_WorldRank == 0)
+    {
+        grbuf.resize(m_WorldSize);
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
```
Comment (JasonRuonanWang) [inline source/adios2/engine/ssc/SscReader.cpp]: You don't need to call MPI_Barrier around MPI collective calls.

File: source/adios2/engine/ssc/SscWriter.cpp [hunk 3]
```diff
@@ -34,6 +36,22 @@ SscWriter::SscWriter(IO &io, const std::string &name, const Mode mode,
     m_GlobalWritePattern.resize(m_WriterSize);
     m_GlobalReadPattern.resize(m_ReaderSize);
 
+    /*
```
Comment (JasonRuonanWang) [inline source/adios2/engine/ssc/SscWriter.cpp]: Can these be removed?

File: source/adios2/engine/ssc/SscWriter.cpp [hunk 4]
```diff
@@ -120,6 +138,233 @@ void SscWriter::SyncMpiPattern()
         std::cout << "SscWriter::SyncMpiPattern, World Rank " << m_WorldRank
                   << ", Writer Rank " << m_WriterRank << std::endl;
     }
+
+
+    std::vector<int> lrbuf;
+    std::vector<int> grbuf;
+    int m_AppID = 0;
```
Comment (JasonRuonanWang) [inline source/adios2/engine/ssc/SscWriter.cpp]: Same as above, these should be in the header file

File: source/adios2/engine/ssc/SscWriter.cpp [hunk 5]
```diff
@@ -120,6 +138,233 @@ void SscWriter::SyncMpiPattern()
...
+    std::vector<std::vector<int>> m_ReaderGlobalMpiInfo;
+
+
+    // Process m_WorldRank == 0 to gather all the local rank m_WriterRank, and find out all the m_WriterRank == 0
+    if (m_WorldRank == 0)
+    {
+        grbuf.resize(m_WorldSize);
+    }
+
+    MPI_Barrier(MPI_COMM_WORLD);
```
Comment (JasonRuonanWang) [inline source/adios2/engine/ssc/SscWriter.cpp]: Barriers are not needed

**Comment by JasonRuonanWang** [comment]
@shevateng0 Thanks for the effort. It looks good but I made some minor comments. You will also need to use clang-format on the files you have changed before they can pass the tests.

File: source/adios2/engine/ssc/SscReader.cpp (hunk 1 above)
Comment (shevateng0) [reply source/adios2/engine/ssc/SscReader.cpp]: I have moved all of the printing codes to the end. They will be printed when m_Verbosity >= 10.

File: source/adios2/engine/ssc/SscReader.cpp [hunk 6]
```diff
@@ -117,6 +117,235 @@ void SscReader::SyncMpiPattern()
         std::cout << "SscReader::SyncMpiPattern, World Rank " << m_WorldRank
                   << ", Reader Rank " << m_ReaderRank << std::endl;
     }
+
+
+    std::vector<int> lrbuf;
+    std::vector<int> grbuf;
+    int m_AppID = 0;
```
[the remark this answers is not in the document]
Comment (shevateng0) [reply source/adios2/engine/ssc/SscReader.cpp]: > These should be put into the header file as class member variables.

I have put them into the header file.

File: source/adios2/engine/ssc/SscWriter.cpp (hunk 3 above)
Comment (shevateng0) [reply source/adios2/engine/ssc/SscWriter.cpp]: > Can these be removed?

All the redundant comments have been removed.

File: source/adios2/engine/ssc/SscWriter.cpp (hunk 4 above)
Comment (shevateng0) [reply source/adios2/engine/ssc/SscWriter.cpp]: > Same as above, these should be in the header file

All the member variables are moved to the header file now.

File: source/adios2/engine/ssc/SscWriter.cpp (hunk 5 above)
Comment (shevateng0) [reply source/adios2/engine/ssc/SscWriter.cpp]: > Barriers are not needed

Thanks. I have deleted those barriers.

**Pull request merged**